peer/network.go

```diff
@@ -265,27 +256,102 @@ func (n *network) AppRequestFailed(_ context.Context, nodeID ids.NodeID, request
+// If [requestID] is not known, this function will emit a log and return a nil error.
+// If the response handler returns an error it is propagated as a fatal error.
+func (n *network) CrossChainAppRequestFailed(ctx context.Context, respondingChainID ids.ID, requestID uint32) error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	log.Debug("received CrossChainAppRequestFailed from chain", "respondingChainID", respondingChainID, "requestID", requestID)
+
+	handler, exists := n.markRequestFulfilled(requestID, true)
+	if !exists {
+		// Should never happen since the engine should be managing outstanding requests
+		log.Error("received CrossChainAppRequestFailed failed to unknown request", "respondingChainID", respondingChainID, "requestID", requestID)
+		return nil
+	}
+
+	return handler.OnFailure(requestID)
+}
+
+// CrossChainAppResponse is invoked when there is a response
+// received from [respondingChainID] regarding a request the VM sent out
+// If [requestID] is not known, this function will emit a log and return a nil error.
+// If the response handler returns an error it is propagated as a fatal error.
+func (n *network) CrossChainAppResponse(ctx context.Context, respondingChainID ids.ID, requestID uint32, response []byte) error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	log.Debug("received CrossChainAppResponse from responding chain", "respondingChainID", respondingChainID, "requestID", requestID)
+
+	handler, exists := n.markRequestFulfilled(requestID, true)
+	if !exists {
+		// Should never happen since the engine should be managing outstanding requests
+		log.Error("received CrossChainAppResponse to unknown request", "respondingChainID", respondingChainID, "requestID", requestID, "responseLen", len(response))
+		return nil
+	}
+
+	return handler.OnResponse(requestID, response)
+}
+
+// markAppRequestFulfilled fetches the handler for [requestID] and marks the request with [requestID] as having been fulfilled.
+// if [isCrossChainRequest] then we do not release the semaphore for [n.activeRequests] as CrossChainRequests have no maximum upperbound.
 // This is called by either [AppResponse] or [AppRequestFailed].
 // assumes that the write lock is held.
-func (n *network) getRequestHandler(requestID uint32) (message.ResponseHandler, bool) {
+func (n *network) markRequestFulfilled(requestID uint32, isCrossChainRequest bool) (message.ResponseHandler, bool) {
 	handler, exists := n.outstandingRequestHandlers[requestID]
 	if !exists {
 		return nil, false
 	}
 	// mark message as processed, release activeRequests slot
 	delete(n.outstandingRequestHandlers, requestID)
-	n.activeRequests.Release(1)
+
+	// [n.activeRequests] controls the maximum number of active outbound requests for all requests that are not CrossChainRequest
+	// If we see a request that is not of type CrossChainRequest, we must release the semaphore
+	// CrossChainRequests have no maximum.
+	if !isCrossChainRequest {
+		n.activeRequests.Release(1)
+	}
```

Comment:
we should probably either comment this on top of `activeRequest` definition or rename `activeAppRequest`

Reply:
refactored this to add `maxActiveCrossChainRequests`